library/src/main/scala/scala/scalajs/js/package.scala

```diff
@@ -105,4 +109,75 @@ package object js {
       "type of a Scala.js library has been called. This is most likely " +
       "because you tried to run Scala.js binaries on the JVM. Make sure you " +
       "are using the JVM version of the libraries.")
+
+  /** Allows to type safely cast an instance of a Scala class to facade trait
+   *
+   *  Use as follows:
+   *  {{{
+   *  js.use(x).as[MyFacade]
+   *  }}}
+   *
+   *  Note that the method calls are only syntactic sugar. There is no overhead
+   *  at runtime for such an operation. Using `use(x).as[T]` is strictly
+   *  equivalent to `x.asInstanceOf[T]` if the compile time check does not fail.
+   *
+   *  == Examples ==
+   *  Given the following facade type:
+   *  {{{
+   *  trait MyFacade extends js.Object {
+   *    def foo(x: Int): String = js.native
+   *    def bar: Int = js.native
+   *  }
+   *  }}}
+   *
+   *  We show a couple of examples:
+   *  {{{
+   *  class MyClass1 {
+   *    @JSExport
+   *    def foo(x: Int): String = x.toString
+   *
+   *    @JSExport
+   *    def bar: Int = 1
+   *  }
+   *
+   *  val x1 = new MyClass1
+   *  js.use(x1).as[MyFacade] // OK
+   *  }}}
+   *
+   *  {{{
+   *  class MyClass2 {
+   *    @JSExport
+   *    def foo(x: Int): String = x.toString
+   *
+   *    @JSExport
+   *    val bar: Int = 1
+   *  }
+   *
+   *  val x2 = new MyClass2
+   *  js.use(x2).as[MyFacade] // OK
+   *  }}}
+   *
+   *  {{{
+   *  class MyClass3 {
+   *    @JSExport
+   *    def foo(x: String): String = x.toString
+   *  }
+   *
+   *  val x3 = new MyClass3
+   *  js.use(x2).as[MyFacade] // Fails: bar is missing and foo has wrong type
+   *  }}}
+   *
+   *  == Restrictions ==
+   *  - Facade types may only be traits and not have any class ancestors
+   *  - Polymorphic methods are currently not supported
+   *  - Facade types defining an apply method cannot used (this is a JavaScript
+   *    restriction).
+   */
+  def use[A](x: A): Using[A] = new Using[A](x)
+
+  /** Helper for syntactic sugar of [[use]]. Do not use directly */
+  final class Using[A](val x: A) extends AnyVal {
```

Comment:
Irrespective of that, `Using` should be in its own file. Classes inside package objects should be avoided (there are still bugs about that in scalac, AFAIK). Having it its own file would also make it easier to override that file and in the entire `macroImpls` package, should we need a completely different implementation (say scala.meta) in a different compiler (say, dotty ^^).